{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import itertools as it\n",
    "import os\n",
    "\n",
    "import collections\n",
    "def recursively_default_dict():\n",
    "    return collections.defaultdict(recursively_default_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demography simulations\n",
    "\n",
    "> adapt demographic information in written format to SLiM recipe.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Files\n",
    "\n",
    "Files used for this simulation.\n",
    "\n",
    "**necessary input**\n",
    "\n",
    "- Table of demographic information. Very sensitive to formatting. \n",
    "> [demos file](https://github.com/SantosJGND/SLiM/blob/master/demos_ABC/demos/PM2013_M4A.txt)\n",
    "   \n",
    "- template recipe, initialize() block.\n",
    "> [template](https://github.com/SantosJGND/SLiM/blob/master/demos_ABC/Recipes/demos_mat/template_matVar.slim)\n",
    "\n",
    "**final product**\n",
    "- [Recipes/demos_mat/M4A0.slim](https://github.com/SantosJGND/SLiM/blob/master/demos_ABC/Recipes/demos_mat/M4A0.slim)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "TabError",
     "evalue": "inconsistent use of tabs and spaces in indentation (ABC_utilities.py, line 575)",
     "output_type": "error",
     "traceback": [
      "Traceback \u001b[1;36m(most recent call last)\u001b[0m:\n",
      "  File \u001b[0;32m\"c:\\users\\floyd\\appdata\\local\\programs\\python\\python37-32\\lib\\site-packages\\IPython\\core\\interactiveshell.py\"\u001b[0m, line \u001b[0;32m3296\u001b[0m, in \u001b[0;35mrun_code\u001b[0m\n    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-3-508448404c1d>\"\u001b[1;36m, line \u001b[1;32m2\u001b[1;36m, in \u001b[1;35m<module>\u001b[1;36m\u001b[0m\n\u001b[1;33m    from tools.ABC_utilities import read_demofile, demos_to_SLiM\u001b[0m\n",
      "\u001b[1;36m  File \u001b[1;32m\"D:\\GitHub\\SLiM\\demos_ABC\\tools\\ABC_utilities.py\"\u001b[1;36m, line \u001b[1;32m575\u001b[0m\n\u001b[1;33m    if M_convert:\u001b[0m\n\u001b[1;37m                 ^\u001b[0m\n\u001b[1;31mTabError\u001b[0m\u001b[1;31m:\u001b[0m inconsistent use of tabs and spaces in indentation\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from tools.ABC_utilities import read_demofile, demos_to_SLiM\n",
    "\n",
    "\n",
    "def demo_to_recipe(demo_file,template,batch= 'test',anc_r= '0',Nsamp= 5,sizes= 500, burnin= 5e4, sim_scale= 1,recipe_dir= 'Recipes/demos_mat/',\n",
    "                    rescale_dict= {},med_samp= False,M_convert= False,directed= False):\n",
    "    \n",
    "    tree, demo_data= read_demofile(demo_file)\n",
    "    \n",
    "    pops, files= demos_to_SLiM(batch, template,tree, demo_data, anc_r= anc_r, Nsamp= Nsamp, sizes= sizes, burnin= burnin, sim_scale= sim_scale,\n",
    "                                                    rescale_dict= rescale_dict,med_samp= med_samp,M_convert= M_convert,directed= directed,\n",
    "                                                    size_key= '\\t{}.setSubpopulationSize({});\\n',\n",
    "                                                    mig_key= '{}.setMigrationRates(c({}), c({}));\\n',\n",
    "                                                    create_key= 'sim.addSubpopSplit(\"{}\", {}, {});\\n')\n",
    "    \n",
    "    \n",
    "    return pops, files\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "batch= 'M4Amedian'\n",
    "recipe_dir= 'Recipes/demos_mat/'\n",
    "\n",
    "\n",
    "demo_file= 'demos/PM2013_M4A.txt'\n",
    "template= 'Recipes/demos_mat/template_matVar.slim'\n",
    "\n",
    "anc_r= '0'\n",
    "Nsamp= 3\n",
    "sizes= 500\n",
    "burnin= 5e4\n",
    "med_samp= True\n",
    "rescale_dict= {\n",
    "    'T': 1 / 20\n",
    "}\n",
    "directed= False\n",
    "M_convert= True\n",
    "\n",
    "pops,files= demo_to_recipe(demo_file, template,batch= batch,anc_r= anc_r,Nsamp= Nsamp, M_convert= M_convert,\n",
    "                           recipe_dir=recipe_dir, med_samp= med_samp,rescale_dict= rescale_dict)\n",
    "\n",
    "\n",
    "#print(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "from scipy.stats import beta\n",
    "\n",
    "def sample_dist_beta(nsample,median,ll_cl,up_cl,blur= 500,assume='norm',func= '',func_args= [3],source= 0):\n",
    "    '''\n",
    "    Use beta distribution to add skew.\n",
    "    '''\n",
    "\n",
    "    if not source: \n",
    "        blur= blur\n",
    "    else:\n",
    "        blur= 500\n",
    "    \n",
    "    window= up_cl - ll_cl\n",
    "    sd_s= (window) / 2\n",
    "    \n",
    "    rate= (median - ll_cl) / window\n",
    "    \n",
    "    t= np.pi / 2\n",
    "    \n",
    "    a= np.sin(rate * t) * blur\n",
    "    b= np.cos(rate * t) * blur\n",
    "    \n",
    "    f= beta.rvs(a, b, size=nsample)\n",
    "    \n",
    "    if not source or up_cl < 1:\n",
    "        f= f * window + ll_cl\n",
    "\n",
    "    if func:\n",
    "        f= [func(x,*func_args) for x in f]\n",
    "    \n",
    "    return f\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Posterior sampling\n",
    "import matplotlib.pyplot as plt\n",
    "#from tools.ABC_utilities import sample_dist_beta\n",
    "\n",
    "fig_dir= 'Figures/demos_samp'\n",
    "os.makedirs(fig_dir, exist_ok=True)\n",
    "fig_dir= fig_dir + '/'\n",
    "\n",
    "\n",
    "demo_file= 'demos/PM2013_M4A.txt'\n",
    "\n",
    "with open(demo_file,'r') as f:\n",
    "    lines= f.readlines()\n",
    "\n",
    "lines= lines[1:]\n",
    "lines= [x.strip().split('\\t') for x in lines]\n",
    "lines= {\n",
    "    ''.join(x[:2]): np.array(x[2:],dtype=float) for x in lines\n",
    "}\n",
    "\n",
    "Nrep= 10000\n",
    "blur= 50\n",
    "samp_dict={}\n",
    "\n",
    "for l,g in lines.items():\n",
    "    \n",
    "    source= 0\n",
    "    if 'M' in l:\n",
    "        source= 1\n",
    "    \n",
    "    t= sample_dist_beta(Nrep,g[0],g[1],g[2],blur= blur,func= '',func_args= [3],source= source)\n",
    "    samp_dict[l]= t\n",
    "    \n",
    "    bins= 50\n",
    "    \n",
    "    plt.figure(figsize=(20,20))\n",
    "    plt.hist(t, bins = bins)\n",
    "    \n",
    "    plt.xticks(fontsize=17)\n",
    "    \n",
    "    plt.xlabel('surface',fontsize= 10)\n",
    "    if source:\n",
    "        plt.xlim(0,1)\n",
    "    else:\n",
    "        plt.xlim(g[1],g[2])\n",
    "    plt.ylabel('density')\n",
    "    plt.title(l)\n",
    "\n",
    "    plt.xticks(fontsize=30)\n",
    "    plt.yticks(fontsize=20)\n",
    "    plt.savefig(fig_dir + 'sampling_dist_{}.png'.format(l),bbox_inches='tight')\n",
    "    plt.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions of interest.\n",
    "\n",
    "### I. Sampling.\n",
    "\n",
    "Currently using sample_dist_beta. This function makes use of the median and confidence interval to determine the skew in the proposed distribution, uses the beta distribution to try to emulate this skew. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import norm\n",
    "\n",
    "def sample_dist(nsample,median,ll_cl,up_cl,assume='norm',func= '',func_args= [3],source= 0):\n",
    "    '''\n",
    "    determine mean and sd from UP-CL 95%.\n",
    "    sample using scipy.\n",
    "    '''\n",
    "    \n",
    "    mean_s= (ll_cl+up_cl) / 2\n",
    "    sd_s= (up_cl - mean_s) / 2\n",
    "    \n",
    "    t= norm.rvs(loc=mean_s,scale= sd_s,size= nsample)\n",
    "    if func:\n",
    "        t= [func(x,*func_args) for x in t]\n",
    "    \n",
    "    return t\n",
    "\n",
    "\n",
    "from scipy.stats import beta\n",
    "\n",
    "\n",
    "\n",
    "def sample_dist_beta(nsample,median,ll_cl,up_cl,blur= 8,assume='norm',func= '',func_args= [3],source= 0):\n",
    "    '''\n",
    "    Use beta distribution to add skew.\n",
    "    '''\n",
    "    \n",
    "    mean_s= (ll_cl+up_cl) / 2\n",
    "    window= up_cl - ll_cl\n",
    "    sd_s= (window) / 2\n",
    "    \n",
    "    rate= (median - ll_cl) / window\n",
    "    t= np.pi / 2\n",
    "\n",
    "    a= np.sin(rate * t) * blur\n",
    "    b= np.cos(rate * t) * blur\n",
    "    \n",
    "    t= beta.rvs(a, b, size=nsample)\n",
    "    \n",
    "    if not source:\n",
    "        t= t * window + ll_cl\n",
    "    \n",
    "    if func:\n",
    "        t= [func(x,*func_args) for x in t]\n",
    "    \n",
    "    return t\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### II. Initialize and sampling blocks\n",
    "\n",
    "The content of these functions deterlines how samples are extracted as well as the initial blocks. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ancestral_initialize(anc_name= 'p1',anc_size= 20000,return_list= True):\n",
    "    anc_intro= '''\n",
    "    sim.addSubpop(\"{}\", {});\n",
    "    c = sim.chromosome;\n",
    "    catn(\"Ancestral: \" + paste(c.ancestralNucleotides(format=\"char\")[0:20],sep=\"\"));\n",
    "    catn();\\n'''\n",
    "    \n",
    "    anc_intro= anc_intro.format(anc_name,str(anc_size))\n",
    "    \n",
    "    anc_intro= \"\"\"1 {\\n\"\"\" + anc_intro  + \"\"\"}\\n\"\"\"\n",
    "    \n",
    "    if return_list:\n",
    "        anc_intro= anc_intro.split('\\n')\n",
    "        anc_intro= [x + '\\n' for x in anc_intro]\n",
    "        \n",
    "    return anc_intro\n",
    "\n",
    "\n",
    "def sample_block(gen= 60000,pops= ['p1','p2'],sizes= [500,500]):\n",
    "    pops= ','.join(pops)\n",
    "    sizes= ','.join([str(x) for x in sizes])\n",
    "    \n",
    "    sample_simple= \"\"\"\n",
    "    g = c();\n",
    "    pops= c({});\n",
    "    samples= c({});\n",
    "    for (x in pops) \n",
    "        g= c(g, sim.subpopulations[x].sampleIndividuals(samples[x]).genomes);\n",
    "\n",
    "    g.outputVCF(vcf_file,simplifyNucleotides=T);\n",
    "    \"\"\"\n",
    "    \n",
    "    sample_simple= sample_simple.format(pops,sizes)\n",
    "    sample_simple= \"\"\"{} late() \"\"\".format(gen) + \"\"\"{\\n\"\"\" + sample_simple\n",
    "    sample_simple= sample_simple.split('\\n')\n",
    "    sample_simple= [x + '\\n' for x in sample_simple]\n",
    "    sample_simple.append( \"\"\"}\\n\"\"\")\n",
    "    \n",
    "    return sample_simple\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
