{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import itertools as it\n",
    "import os\n",
    "\n",
    "import collections\n",
    "def recursively_default_dict():\n",
    "    return collections.defaultdict(recursively_default_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demography simulations\n",
    "\n",
    "> adapt demographic information in written format to SLiM recipe.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def str_simple(string,tag='(',end= ')'):\n",
    "    '''\n",
    "    return index of first comma in the first layer of a nested tuple string.\n",
    "    '''\n",
    "    left= 0\n",
    "    right= 0\n",
    "    \n",
    "    for s in range(1,len(string)):\n",
    "        \n",
    "        if string[s] == tag:\n",
    "            left += 1\n",
    "        if string[s] == end:\n",
    "            right += 1\n",
    "        \n",
    "        if left == right:\n",
    "            return s + 1\n",
    "\n",
    "\n",
    "def split_middle_comma(tree_sr):\n",
    "    '''\n",
    "    split a 3-element tuple string.\n",
    "    '''\n",
    "    line= str(tree_sr)\n",
    "    \n",
    "    if line[1] != '(':\n",
    "        #\n",
    "        line= line.strip(')').strip('(')\n",
    "        line= line.split(',')\n",
    "        return line\n",
    "    \n",
    "    indx_split= str_simple(tree_sr)\n",
    "    \n",
    "    pos1= tree_sr[:indx_split][1:]\n",
    "    pos2= tree_sr[indx_split+1:][:-1]\n",
    "    \n",
    "    if pos2[0] != '(':\n",
    "        pos2,pos2b= pos2.split(',')\n",
    "        return [pos1,pos2,pos2b]\n",
    "    \n",
    "    \n",
    "    pos2= '({})'.format(pos2)\n",
    "    \n",
    "    pos2_idx= str_simple(pos2)\n",
    "    \n",
    "    pos2b= pos2[:pos2_idx][1:]\n",
    "    pos2c= pos2[pos2_idx+1:][:-1]\n",
    "    \n",
    "    return [pos1,pos2b,pos2c]\n",
    "\n",
    "\n",
    "\n",
    "def det_parenthesis(node_str,det= '('):\n",
    "    if node_str[0]== det:\n",
    "        return tuple_tree_rec(node_str)\n",
    "    else:\n",
    "        return node_str\n",
    "\n",
    "    \n",
    "def tuple_tree_rec(line):\n",
    "    '''\n",
    "    read binary trees in ((node1,node2),node3). \n",
    "    Recursive = inner branch number limited by machine recursion limit.\n",
    "    \n",
    "    '''\n",
    "    test= split_middle_comma(line)\n",
    "    \n",
    "    return {\n",
    "        test[0]: det_parenthesis(test[0]),\n",
    "        test[1]: det_parenthesis(test[1]),\n",
    "        'node': test[2]\n",
    "    }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## read dems file\n",
    "\n",
    "def read_demofile(filename):\n",
    "    '''\n",
    "    read demographic history table. Simple.\n",
    "    first line must describe format in ((n1,n2),n3) format.\n",
    "    1st column: T,N or M for time, size and migration.\n",
    "    2nd colomn: ordered times, population names and concatenated (into tuples) population names, respectively.\n",
    "    Further columns are values. 3 = mean or sd, 2-4 = lower & upper confidence intervals respectively.\n",
    "    \n",
    "    return tree from first line, dictionary of values w/ levels 1st then second columns.\n",
    "    '''\n",
    "    args= recursively_default_dict()\n",
    "    \n",
    "    with open(filename,'r') as fp:\n",
    "        lines= fp.readlines()\n",
    "    \n",
    "    tree= lines[0].strip()\n",
    "    tree= tuple_tree_rec(tree)\n",
    "    \n",
    "    mig_list= []\n",
    "    for line in lines[1:]:\n",
    "        line= line.split()\n",
    "        if line[0] != 'M':\n",
    "            args[line[0]][line[1]]= np.array(line[2:],dtype= int)\n",
    "        else: \n",
    "            mig_list.append(line[1:])\n",
    "        \n",
    "    \n",
    "    M_dict= recursively_default_dict()\n",
    "    \n",
    "    for line in mig_list:\n",
    "        source= line[0].split('-')[0]\n",
    "        sink= line[0].split('-')[1]\n",
    "        \n",
    "        M_dict[source][sink]= np.array(line[1:],dtype= float)\n",
    "        \n",
    "    args['M']= M_dict\n",
    "    \n",
    "    return tree, args\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_unzip(str_key):\n",
    "    '''\n",
    "    return elements of tuple.\n",
    "    '''\n",
    "    sp_idx= str_simple(str_key)\n",
    "    sp1= str_key[1:sp_idx]\n",
    "    \n",
    "    sp2= '(' + str_key[sp_idx+1:]\n",
    "    \n",
    "    sp2_idx= str_simple(sp2)\n",
    "    sp2= sp2[1:sp2_idx]\n",
    "    \n",
    "    return sp1,sp2\n",
    "    \n",
    "def remove_node(str_key):\n",
    "    '''\n",
    "    remove 3d position from tuple.\n",
    "    '''\n",
    "    \n",
    "    if str_key[0] != '(':\n",
    "        return str_key\n",
    "    \n",
    "    sp1, sp2= split_unzip(str_key)\n",
    "    \n",
    "    return '({})'.format(sp1 + ','+ sp2)\n",
    "\n",
    "\n",
    "def pops_rec_decide(str_key,pops= []):\n",
    "    '''\n",
    "    co-factor to get_pops. If no enclosure, append to listn else launch get_pops within the next level.\n",
    "    '''\n",
    "    if str_key[0] != '(':\n",
    "        pops.append(str_key)\n",
    "    else:\n",
    "        get_pops(str_key,pops= pops)\n",
    "\n",
    "def get_pops(str_key,pops= []):\n",
    "    '''\n",
    "    Extract pops from tuples ignoring 3d position. Allows for nested loops.\n",
    "    return list of pops. returns string if not enclosed in parentheses.\n",
    "    '''\n",
    "    pps= remove_node(str_key)\n",
    "    \n",
    "    if str_key[0] != '(':\n",
    "        pops.append(str_key)\n",
    "        if ',' in str_key:\n",
    "            pops.extend(str_key.split(','))\n",
    "    else:\n",
    "        pps= split_unzip(str_key)\n",
    "        \n",
    "        for p in pps:\n",
    "            pops_rec_decide(p,pops= pops)\n",
    "    \n",
    "    return pops\n",
    "\n",
    "\n",
    "def make_pop(str_key):\n",
    "    '''\n",
    "    make pop name as concatenation of pops in nested tuple.\n",
    "    '''\n",
    "    pop_name= get_pops(str_key,pops= [])\n",
    "    #print('kj: {}'.format(pop_name))\n",
    "    pop_name= 'p' + ''.join(pop_name)\n",
    "    return pop_name\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def decide_rec_list(key_str,tree_dict,tree_summ,times,demo_data,func,int_sizes= [],demo_tape= []):\n",
    "    \n",
    "    if key_str[0] == '(':\n",
    "        tree_fill_list(times,tree_dict[key_str],demo_data,tree_summ,int_sizes= int_sizes,demo_tape=demo_tape)\n",
    "    \n",
    "\n",
    "def tree_fill_list(times,tree_dict, demo_data, tree_summ,int_sizes= [], demo_tape= []):\n",
    "    '''\n",
    "    returns list of tree node information.\n",
    "        - 'T': node index in time array. \n",
    "        - 'M': migratory events concerning new populations.\n",
    "        - 'N': create population events. \n",
    "    \n",
    "    '''\n",
    "    #print(demo_tape)\n",
    "    tree_nodes= {\n",
    "        x: remove_node(x) for x in tree_dict.keys() if x != 'node'\n",
    "    }\n",
    "    \n",
    "    nod_t= int(tree_dict['node'])\n",
    "    nod_t= times[nod_t]\n",
    "    \n",
    "    node_mig= {}\n",
    "    node_n= {}\n",
    "    \n",
    "    gnode_sort= {}\n",
    "    \n",
    "    for k,g in tree_nodes.items():\n",
    "        \n",
    "        if g[0] != '(':\n",
    "            node_n[g]= demo_data['N'][g]\n",
    "            if g in demo_data['M'].keys():\n",
    "                \n",
    "                node_mig[make_pop(k)]= {\n",
    "                    make_pop(k): f for k,f in demo_data['M'][g].items()\n",
    "                }\n",
    "        \n",
    "        if g[0] == '(':\n",
    "            gnode= tree_dict[k]['node']\n",
    "            gnode_sort[g]= gnode\n",
    "    \n",
    "    if gnode_sort:\n",
    "        ## only using the keys, but need them sorted by values.\n",
    "        gnode_sorted= sorted(gnode_sort,key= gnode_sort.get)\n",
    "        \n",
    "        ## from root to leaves: T1-N decreases, Node number increases.\n",
    "        startat= int(tree_dict['node']) - 2\n",
    "        \n",
    "        ## gnode times is used to identify key in demos table, where internal branch names are coded T1-TN\n",
    "        gnode_t= {\n",
    "            gnode_sorted[x]:int_sizes[startat - x] for x in range(len(gnode_sorted))\n",
    "        }\n",
    "        \n",
    "        gnode_sizes= {\n",
    "            v:demo_data['N'][g] for v,g in gnode_t.items()\n",
    "        }\n",
    "        \n",
    "        gnode_parent= {}\n",
    "        \n",
    "        for gnode in gnode_sorted:\n",
    "            ## parents are not used. \n",
    "            parent= [x[0] for x in tree_summ['edges'] if gnode == x[1]][0]\n",
    "            parent= make_pop(parent)\n",
    "            gnode_parent[gnode]= parent\n",
    "            \n",
    "            gnode_pops= get_pops(gnode,pops= [])\n",
    "            \n",
    "            gnode_pops= ','.join(gnode_pops)\n",
    "            \n",
    "            if gnode_pops in demo_data['M'].keys():\n",
    "                if gnode_pops == '1,2':\n",
    "                node_mig[make_pop(gnode)]= {\n",
    "                    make_pop(f): x for f,x in demo_data['M'][gnode_pops].items()\n",
    "                }\n",
    "            \n",
    "        \n",
    "        gnode_N= {\n",
    "            g: gnode_sizes[g] for g in gnode_sorted\n",
    "        }\n",
    "        \n",
    "        node_n.update(gnode_N)\n",
    "    \n",
    "    demo_tape.append({\n",
    "        'node': nod_t,\n",
    "        'T': demo_data['T'][nod_t],\n",
    "        'N': node_n,\n",
    "        'M': node_mig,\n",
    "    \n",
    "    })\n",
    "    \n",
    "    for edge in tree_nodes.keys():\n",
    "        decide_rec_list(edge,tree_dict,tree_summ,times,demo_data,tree_fill_list,\n",
    "                        int_sizes= int_sizes,demo_tape=demo_tape)\n",
    "    \n",
    "    return demo_tape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tree_nodes(tree,nodes=[],edges= [],leaves= []):\n",
    "    '''\n",
    "    break down binary dictionary tree.\n",
    "    '''\n",
    "\n",
    "    \n",
    "    for node in tree.keys():\n",
    "        if node == 'node':\n",
    "            continue\n",
    "        nodes.append(node)\n",
    "        \n",
    "        if tree[node]== node:\n",
    "            leaves.append(node)\n",
    "            continue\n",
    "        \n",
    "        for dt in tree[node].keys():\n",
    "            if dt != 'node':\n",
    "                edges.append((remove_node(node),remove_node(dt)))\n",
    "        \n",
    "        get_tree_nodes(tree[node],nodes= nodes,edges= edges,leaves= leaves)\n",
    "    \n",
    "    summary_dict= {\n",
    "        'nodes': nodes,\n",
    "        'edges': edges,\n",
    "        'leaves': leaves\n",
    "    }\n",
    "    \n",
    "    return summary_dict\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import norm\n",
    "\n",
    "def sample_dist(nsample,median,ll_cl,up_cl,assume='norm',func= '',func_args= [3],source= 0):\n",
    "    '''\n",
    "    determine mean and sd from UP-CL 95%.\n",
    "    sample using scipy.\n",
    "    '''\n",
    "    \n",
    "    mean_s= (ll_cl+up_cl) / 2\n",
    "    sd_s= (up_cl - mean_s) / 2\n",
    "    \n",
    "    t= norm.rvs(loc=mean_s,scale= sd_s,size= nsample)\n",
    "    if func:\n",
    "        t= [func(x,*func_args) for x in t]\n",
    "    \n",
    "    return t\n",
    "\n",
    "\n",
    "from scipy.stats import beta\n",
    "\n",
    "\n",
    "\n",
    "def sample_dist_beta(nsample,median,ll_cl,up_cl,blur= 3,assume='norm',func= '',func_args= [3],source= 0):\n",
    "    '''\n",
    "    Use beta distribution to add skew.\n",
    "    '''\n",
    "    \n",
    "    mean_s= (ll_cl+up_cl) / 2\n",
    "    window= up_cl - mean_s\n",
    "    sd_s= (window) / 2\n",
    "    \n",
    "    rate= (median - ll_cl) / window\n",
    "    t= np.pi / 2\n",
    "\n",
    "    a= np.sin(rate * t) * blur\n",
    "    b= np.cos(rate * t) * blur\n",
    "    \n",
    "    t= beta.rvs(a, b, size=nsample)\n",
    "    \n",
    "    if not source:\n",
    "        t= t * window + ll_cl\n",
    "    \n",
    "    if func:\n",
    "        t= [func(x,*func_args) for x in t]\n",
    "    \n",
    "    return t\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rep_chose(val,g,tree_demo,assume= 'norm',func= int,sample_func= sample_dist_beta,source= 0,track= ['M']):\n",
    "    \n",
    "    branch= source\n",
    "    \n",
    "    if val in track:\n",
    "        branch= val\n",
    "    \n",
    "    if isinstance(g, dict):\n",
    "        return return_replica(tree_demo[val],assume= assume,source= branch)\n",
    "    \n",
    "    if type(g).__module__ == np.__name__:\n",
    "        g= sample_func(1,*g,func= round,source= source)[0]\n",
    "        if func: \n",
    "            g= func(g)\n",
    "    \n",
    "    elif isinstance(g,list):\n",
    "        g= [g[0],sample_func(1,*g[1],func= round,func_args= [],source= source)[0]]\n",
    "\n",
    "    return g\n",
    "\n",
    "\n",
    "\n",
    "def return_replica(tree_demo,assume='norm',func= float,sample_func= sample_dist_beta,source= 0):\n",
    "    '''\n",
    "    sample topologie from from probab demos dictionary. \n",
    "    '''\n",
    "    return {\n",
    "        val: rep_chose(val,g,tree_demo,assume= assume,func= func,sample_func= sample_func,source= source) for val,g in tree_demo.items()\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ancestral_initialize(anc_name= 'p1',anc_size= 20000,return_list= True):\n",
    "    anc_intro= '''\n",
    "    sim.addSubpop(\"{}\", {});\n",
    "    c = sim.chromosome;\n",
    "    catn(\"Ancestral: \" + paste(c.ancestralNucleotides(format=\"char\")[0:20],sep=\"\"));\n",
    "    catn();\\n'''\n",
    "    \n",
    "    anc_intro= anc_intro.format(anc_name,str(anc_size))\n",
    "    \n",
    "    anc_intro= \"\"\"1 {\\n\"\"\" + anc_intro  + \"\"\"}\\n\"\"\"\n",
    "    \n",
    "    if return_list:\n",
    "        anc_intro= anc_intro.split('\\n')\n",
    "        anc_intro= [x + '\\n' for x in anc_intro]\n",
    "        \n",
    "    return anc_intro\n",
    "\n",
    "\n",
    "def sample_block(gen= 60000,pops= ['p1','p2'],sizes= [500,500]):\n",
    "    pops= ','.join(pops)\n",
    "    sizes= ','.join([str(x) for x in sizes])\n",
    "    \n",
    "    sample_simple= \"\"\"\n",
    "    g = c();\n",
    "    pops= c({});\n",
    "    samples= c({});\n",
    "    for (x in pops) \n",
    "        g= c(g, sim.subpopulations[x].sampleIndividuals(samples[x]).genomes);\n",
    "\n",
    "    g.outputVCF(vcf_file,simplifyNucleotides=T);\n",
    "    \"\"\"\n",
    "    \n",
    "    sample_simple= sample_simple.format(pops,sizes)\n",
    "    sample_simple= \"\"\"{} late() \"\"\".format(gen) + \"\"\"{\\n\"\"\" + sample_simple\n",
    "    sample_simple= sample_simple.split('\\n')\n",
    "    sample_simple= [x + '\\n' for x in sample_simple]\n",
    "    sample_simple.append( \"\"\"}\\n\"\"\")\n",
    "    return sample_simple\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def demos_to_SLiM(batch, recipe_dir, template, tree, demo_data, anc_r= 'anc', Nsamp= 5, sizes= 500, burnin= 5e4,size_key= '\\t{}.setSubpopulationSize({});\\n',\n",
    "                                                    sample_func= sample_dist_beta, mig_key= '{}.setMigrationRates(c({}), c({}));\\n',\n",
    "                                                    create_key= 'sim.addSubpopSplit(\"{}\", {}, {});\\n'):\n",
    "\n",
    "    ###\n",
    "    ###\n",
    "    tree_summ= get_tree_nodes({anc_r:tree},nodes=[],edges= [],leaves= [])\n",
    "\n",
    "    #\n",
    "    anc_name= anc_r\n",
    "    anc_size= demo_data['N'][anc_r]\n",
    "    print(anc_size)\n",
    "    times_order= [0] + sorted(demo_data['T'].keys(),reverse= True)\n",
    "    int_sizes= [x for x in demo_data['N'].keys() if x[0] == 'T']\n",
    "    int_sizes= sorted(int_sizes)\n",
    "\n",
    "    tree_demo= tree_fill_list(times_order,tree, demo_data, tree_summ, int_sizes= int_sizes,demo_tape= [])\n",
    "    ###\n",
    "    ###\n",
    "\n",
    "    template_file= recipe_dir + template\n",
    "    \n",
    "    with open(template_file,'r') as fp:\n",
    "        template= fp.readlines()\n",
    "\n",
    "    pops= tree_summ['leaves']\n",
    "    pops= [make_pop(x) for x in pops]\n",
    "    sizes= [sizes] * len(pops)\n",
    "\n",
    "    files= []\n",
    "    anc_sample= sample_func(Nsamp,*anc_size,func= int,func_args= [])\n",
    "    \n",
    "    for idx_sim in range(Nsamp):\n",
    "        temp= list(template)\n",
    "\n",
    "        ## get data\n",
    "        asize= anc_sample[idx_sim]\n",
    "        \n",
    "        replic= [return_replica(x,sample_func= sample_func) for x in tree_demo]\n",
    "        replic= {x['node']: x for x in replic}\n",
    "        \n",
    "\n",
    "        ## existing pops, to decide whether to update or create.\n",
    "        existing_pops= [anc_name]\n",
    "        trail_migs= []\n",
    "\n",
    "        time_spent= int(burnin)\n",
    "        ## setup ancestral init. \n",
    "        anc_lines= ancestral_initialize(anc_name= make_pop(anc_name),anc_size= asize)\n",
    "\n",
    "        temp.extend(anc_lines)\n",
    "\n",
    "        node_order= sorted(replic.keys(),reverse= True)\n",
    "\n",
    "        for nod in node_order: \n",
    "            new_lines= ['\\n']\n",
    "\n",
    "            Tev= replic[nod]['T']\n",
    "\n",
    "            new_lines.append(str(time_spent) + ' {\\n')\n",
    "            time_spent += int(Tev)\n",
    "\n",
    "            deprecate= []\n",
    "\n",
    "            for br in replic[nod]['N'].keys():            \n",
    "                pop_name= make_pop(br)\n",
    "                #print(pop_name)\n",
    "                parent= [x[0] for x in tree_summ['edges'] if br == x[1]][0]\n",
    "                #print(parent)\n",
    "                parent= make_pop(parent)\n",
    "                pop_size= int(replic[nod]['N'][br])\n",
    "\n",
    "                if pop_name in existing_pops:\n",
    "                    # change size\n",
    "                    size_change= '\\t' + size_key.format(pop_name,pop_size)\n",
    "                    new_lines.append(size_change)\n",
    "\n",
    "                else:\n",
    "                    # create population. \n",
    "                    create_pop= '\\t' + create_key.format(pop_name,pop_size,parent) ###\n",
    "                    new_lines.append(create_pop)\n",
    "                    existing_pops.append(pop_name)\n",
    "                    deprecate.append(parent)\n",
    "\n",
    "\n",
    "            ### remove old populations\n",
    "            deprecate= list(set(deprecate))\n",
    "\n",
    "            for old in deprecate:\n",
    "                size_change= size_key.format(old,0)\n",
    "                new_lines.append(size_change)\n",
    "\n",
    "            ### migration\n",
    "            new_trail=[]\n",
    "            for trail in trail_migs:\n",
    "                if trail[0] in existing_pops and trail[1] in existing_pops:\n",
    "                    mig_change= '\\t' + mig_key.format(trail[0],trail[1],trail[2])\n",
    "                    new_lines.append(mig_change)\n",
    "                else:\n",
    "                    new_trail.append(trail)\n",
    "\n",
    "            trail_migs= new_trail\n",
    "\n",
    "            for mr in replic[nod]['M'].keys():\n",
    "                # set migration rate.\n",
    "                for dr,v in replic[nod]['M'][mr].items():\n",
    "\n",
    "                    if mr in existing_pops and dr in existing_pops:\n",
    "                        mig_change= '\\t' + mig_key.format(mr,dr,v)\n",
    "                        new_lines.append(mig_change)\n",
    "\n",
    "                    else:\n",
    "                        trail_migs.append((mr,dr,v))\n",
    "\n",
    "            new_lines.append('}\\n\\n')\n",
    "\n",
    "            temp.extend(new_lines)\n",
    "\n",
    "        ## sample: \n",
    "\n",
    "        sample_simple= sample_block(time_spent,pops,sizes)\n",
    "        temp.extend(sample_simple)\n",
    "\n",
    "        #### write new recipe\n",
    "        recipe_name= '{}{}.slim'.format(batch,idx_sim)\n",
    "        recipe_name= recipe_dir + recipe_name\n",
    "        files.append(recipe_name)\n",
    "        \n",
    "        with open(recipe_name,'w') as fp:\n",
    "            fp.write(''.join(temp))\n",
    "        \n",
    "        \n",
    "    \n",
    "    \n",
    "    pops= [make_pop(x) for x in tree_summ['leaves']]\n",
    "    \n",
    "    return pops, files\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[10543  1234 71162]\n",
      "{'((1,2,3),3,2)': '((1,2,3),3)', '4': '4'}\n",
      "{'(1,2,3)': '(1,2)', '3': '3'}\n",
      "hi\n",
      "{'1': '1', '2': '2'}\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def demo_to_recipe(demo_file,template,batch= 'test',anc_r= '0',Nsamp= 5,sizes= 500, burnin= 5e4, recipe_dir= 'Recipes/demos_mat/'):\n",
    "    \n",
    "    tree, demo_data= read_demofile(demo_file)\n",
    "    #print(demo_data['N'])\n",
    "    pops, files= demos_to_SLiM(batch, recipe_dir, template,tree, demo_data, anc_r= anc_r, Nsamp= Nsamp, sizes= sizes, burnin= burnin,\n",
    "                                                    size_key= '\\t{}.setSubpopulationSize({});\\n',\n",
    "                                                    mig_key= '{}.setMigrationRates(c({}), c({}));\\n',\n",
    "                                                    create_key= 'sim.addSubpopSplit(\"{}\", {}, {});\\n')\n",
    "    \n",
    "    \n",
    "    return pops, files\n",
    "\n",
    "\n",
    "\n",
    "batch= 'M4A'\n",
    "recipe_dir= 'Recipes/demos_mat/'\n",
    "\n",
    "\n",
    "demo_file= 'Prado&Martinez2013_M4A.txt'\n",
    "demo_file= 'demos/PM2013_M3.txt'\n",
    "template= 'template_matVar.slim'\n",
    "\n",
    "anc_r= '0'\n",
    "Nsamp= 5\n",
    "sizes= 500\n",
    "burnin= 5e4\n",
    "\n",
    "\n",
    "pops,files= demo_to_recipe(demo_file, template,batch= batch,anc_r= anc_r,Nsamp= Nsamp, recipe_dir=recipe_dir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Recipes/demos_mat/M4A0.slim',\n",
       " 'Recipes/demos_mat/M4A1.slim',\n",
       " 'Recipes/demos_mat/M4A2.slim',\n",
       " 'Recipes/demos_mat/M4A3.slim',\n",
       " 'Recipes/demos_mat/M4A4.slim']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1': 0}"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "t= {\"1\":0,\n",
    "   \"3\":0}\n",
    "\n",
    "t.pop('3')\n",
    "t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
